### DISCUSSION TITLE
add a request body to a path

### DISCUSSION THREAD

**Discussion by jmls** [pb33f/libopenapi#345] [Q&A]
I have an openApi 3,0 json spec file. Unfortunately one of the paths is missing the  request body, so I thought I'd go on a voyage of discovery , learn go and try and use this library to add it ;) 

so far, I've managed to read in the json spec, build the v3 model , find the path and get the Post . I was also able to add a "basic" requestBody 

```
        doc, _ := libopenapi.NewDocument(readStdin())
        api, _ := doc.BuildV3Model()
        
	cubbyholePost := api.Model.Paths.PathItems.GetOrZero("/cubbyhole/{path}").Post

	cubbyholePost.RequestBody = &v3.RequestBody{
		Description: "this is a new request body",
		Required:    openapi3.BoolPtr(true),

	}

	rend, _ := cubbyholePost.Render()

	fmt.Print(string(rend))
```
and I do indeed get  expected data

```
requestBody:
    description: this is a new request body
    required: true
operationId: "cubbyhole-write"
responses:
    "200":
        description: "OK"
summary: "Store a secret at the specified location."
tags:
    - "secrets"
```

however, I've now come to a screeching halt.  What I want to do is to add a content property  similar to this

```
"content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/CubbyholeWriteRequestBody"
              }
            }
          },
```    
and to  define a model called CubbyholeWriteRequestBody with various properties into the components/schemas tree

some hints and or pointers would be much appreciated. 

It may be because of my go noobness that I'm missing something really obvious

**Comment by jmls** [comment]
yeah unfortunately still struggling here. Thought I'd try and also add a parameter the "yml" way

```
				yml := `description: id id of identity provider
						required: true
						deprecated: false
						name: id
						in: path
						allowEmptyValue: false
						schema:
						type: string`

				var idxNode yaml.Node
				var n v3.Parameter

				n.Name = "id"

				yaml.Unmarshal([]byte(yml), &idxNode)

				low.BuildModel(idxNode.Content[0], &n)

				operation.Parameters = append(operation.Parameters, v3.NewParameter(n.GoLow()))

```
compiles but get segmentation violation

I'm not sure I'm getting a grip on the whole log/high/porcelain thing

**Reply by daveshanley** [reply]
Hi, 

Did you see this yet?

https://pb33f.io/libopenapi/modifying/#using-references

**Reply by daveshanley** [reply]
Everything you're trying to do, you should not need to touch the low level, you only ever need the low level, when you're building tools using the library, if you're just using models, then you only need the porcelain layer. 

All the docs for things you want to do are here: 

https://pb33f.io/libopenapi/modifying/

**Reply by jmls** [reply]
many thanks ! . will investigate and work it out

**Comment by jmls** [comment]
I've made some great progress based on the help given so far . At least, I think I have. :) 

Would appreciate some help on  a couple of things. In the following code, I want to add a response to an operation. This operation can return one of two models depending on a query parameter (yuck, not my spec!) . I am stuck on 

#1 how to define a schema of type array
#2 how to define the response as one of two schemas

```
func fixCubbyholeGet(api *libopenapi.DocumentModel[v3.Document]) {
	readMap := orderedmap.New[string, *base.SchemaProxy]()

	readMap.Set("CubbyholeReadResponse", base.CreateSchemaProxy(&base.Schema{
		Type: []string{"object"},
	}))

	readSchema := base.CreateSchemaProxy(&base.Schema{
		Type:       []string{"object"},
		Properties: readMap,
	})

	api.Model.Components.Schemas.Set("CubbyholeReadResponse", readSchema)

	listMap := orderedmap.New[string, *base.SchemaProxy]()

	listMap.Set("CubbyholeListResponse", base.CreateSchemaProxy(&base.Schema{
		Type: []string{"object"},
		/* #1 how do I maker this into a response that is
		 * { "keys": []}
		 * where keys is an array of strings
		 */

	}))

	listSchema := base.CreateSchemaProxy(&base.Schema{
		Type:       []string{"object"},
		Properties: listMap,
	})

	api.Model.Components.Schemas.Set("CubbyholeReadResponse", readSchema)
	api.Model.Components.Schemas.Set("CubbyholeListResponse", listSchema)

	okResponseContentMap := orderedmap.New[string, *v3.MediaType]()
	okResponseContentMap.Set("application/json", &v3.MediaType{
		Schema: base.CreateSchemaProxyRef("#/components/schemas/CubbyholeReadResponse"),
	})

	cubbyholeGet := api.Model.Paths.PathItems.GetOrZero("/cubbyhole/{path}").Get

	response, _ := cubbyholeGet.Responses.Codes.Get("200")

	response.Content = okResponseContentMap
	// #2 how do I set the response to be one of either cubbyHoleListSchema or cubbyHoleReadSchema

}
```

this is the code so far. It generates a response schema for "CubbyholeReadResponse"  so everything works except for #1 and #2
